File: perf/benchmarks/apps/crc32/crc32.py
# crc32 — CRC32 of a deterministic byte stream application benchmark.
#
# Generate N bytes via LCG, then compute CRC32 (IEEE 802.3 polynomial) byte
# by byte using a precomputed 256-entry table built from the same polynomial.
# Argument: N (default 5_000_000).

import sys
import time


def build_table() -> list[int]:
    table = [0] * 256
    for n in range(256):
        c = n
        for _ in range(8):
            c = (c >> 1) ^ (0xEDB88320 if c & 1 else 0)
        table[n] = c
    return table


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 5_000_000
    src = bytearray(n)
    seed = 1
    for i in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        src[i] = (seed >> 8) & 0xFF
    table = build_table()

    t0 = time.perf_counter_ns()
    crc = 0xFFFFFFFF
    for b in src:
        crc = table[(crc ^ b) & 0xFF] ^ (crc >> 8)
    crc ^= 0xFFFFFFFF
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{crc}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/apps/crc32/crc32.cpp
// crc32 — CRC32 of a deterministic byte stream application benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

static void buildTable(uint32_t* t) {
    for (uint32_t n = 0; n < 256; ++n) {
        uint32_t c = n;
        for (int k = 0; k < 8; ++k) {
            c = (c >> 1) ^ ((c & 1) ? 0xEDB88320u : 0u);
        }
        t[n] = c;
    }
}

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 5000000;
    std::vector<uint8_t> src(n);
    int64_t seed = 1;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        src[i] = static_cast<uint8_t>((seed >> 8) & 0xFF);
    }
    uint32_t table[256];
    buildTable(table);

    auto t0 = std::chrono::steady_clock::now();
    uint32_t crc = 0xFFFFFFFFu;
    for (int64_t i = 0; i < n; ++i) {
        crc = table[(crc ^ src[i]) & 0xFF] ^ (crc >> 8);
    }
    crc ^= 0xFFFFFFFFu;
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%u\n", crc);
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
